Mathlib/Data/List/Chain.lean

```diff
@@ -304,6 +304,28 @@ theorem Chain'.imp_head {x y} (h : ∀ {z}, R x z → R y z) {l} (hl : Chain' R
     Chain' R (y :: l) :=
   hl.tail.cons' fun _ hz => h <| hl.rel_head? hz
 
+theorem chain'_getElem (h : List.Chain' R l) (n : ℕ) (h' : n + 1 < l.length) :
+    R l[n] l[n+1] :=
+  chain'_pair.mp <| Chain'.infix h ⟨l.take n, l.drop (n + 2), by simp⟩
+
+theorem chain'_of_not (h : ¬List.Chain' R l) : ∃ n : ℕ, ∃ h : n + 1 < l.length, ¬R l[n] l[n+1] := by
+  contrapose! h
+  induction l with
+  | nil => simp
+  | cons head tail ih =>
+      rw [chain'_cons']
+      constructor
+      · by_cases h' : tail.length = 0
+        · simp [eq_nil_iff_length_eq_zero.mpr h']
+        · intro y yh
+          simp only [head?_eq_getElem?, Option.mem_def] at yh
+          obtain ⟨_, rfl⟩ := getElem?_eq_some_iff.mp yh
+          have := h 0 (by rw [length_cons]; omega)
+          rwa [getElem_cons_zero] at this
+      · refine ih (fun n h' => ?_)
```

Comment:
Mathlib style slightly prefers \mapsto instead of =>. (I wouldn't insist on this.)